src/plugins/intel_cpu/src/utils/serialize.cpp

```diff
@@ -0,0 +1,190 @@
+        std::string name = nodes_it->attribute("name").value();
+        if (name.empty())
+            continue;
+        auto result = model->output(i).get_node_shared_ptr();
+        ov::descriptor::set_ov_tensor_legacy_name(result->input_value(0).get_tensor(), name);
+    }
+}
+
+void ModelDeserializer::operator>>(std::shared_ptr<ov::Model>& model) {
+    if (auto mmap_stream = dynamic_cast<MmapStream*>(&m_istream)) {
+        process_mmap(model, mmap_stream->m_memory);
+    } else {
+        process_stream(model);
+    }
+}
+
+void ModelDeserializer::process_mmap(std::shared_ptr<ov::Model>& model,
+                                     const std::shared_ptr<ov::MappedMemory>& mmemory) {
+    // Note: Don't use seekg with mmaped stream. This may affect the performance of some models.
+    // Get file size before seek content.
+    // Blob from cache may have other header, so need to skip this.
+    auto buffer_base = mmemory->data();
+    const auto file_size = mmemory->size();
+    const size_t hdr_pos = m_istream.tellg();
+
+    pass::StreamSerialize::DataHeader hdr = {};
+    std::memcpy(reinterpret_cast<char*>(&hdr), buffer_base + hdr_pos, sizeof hdr);
+
+    // Check if model header contains valid data.
+    bool is_valid_model = (hdr.custom_data_offset == sizeof(hdr) + hdr_pos) &&
+                          (hdr.custom_data_size == hdr.consts_offset - hdr.custom_data_offset) &&
+                          (hdr.consts_size == hdr.model_offset - hdr.consts_offset) &&
+                          (hdr.model_size = file_size - hdr.model_offset);
+    if (!is_valid_model) {
+        OPENVINO_THROW("[CPU] Could not deserialize by device xml header.");
+    }
+
+    // Read model input/output precisions.
+    pugi::xml_document xml_in_out_doc;
+    if (hdr.custom_data_size > 0lu) {
+        auto res = xml_in_out_doc.load_buffer(buffer_base + hdr.custom_data_offset, hdr.custom_data_size, pugi::parse_default, pugi::encoding_utf8);
+        if (res.status != pugi::status_ok) {
+            OPENVINO_THROW("[CPU] Could to deserialize custom data.");
+        }
+    }
+
+    // Map blob content
+    std::shared_ptr<ov::AlignedBuffer> weights_buf;
+    if (hdr.consts_size) {
+        weights_buf = std::make_shared<ov::SharedBuffer<std::shared_ptr<MappedMemory>>>(buffer_base + hdr.consts_offset,
+                                                                                        hdr.consts_size,
+                                                                                        mmemory);
+    }
+
+    // XML content
+    auto xml_buff = std::make_shared<std::string>();
+    if (m_cache_decrypt) {
+        std::string (*const* dec_ptr)(const std::string&) = m_cache_decrypt.target<std::string(*)(const std::string&)>();
+        if (dec_ptr && *dec_ptr == codec_xor_str) {
+            xml_buff->reserve(hdr.model_size + 1);
+            codec_xor(&((*xml_buff)[0]), buffer_base + hdr.model_offset, hdr.model_size);
+        } else {
```

Comment:
This solution is too implicit and not that easy to understand. Also it's hard to extend. I think we can do better.
1. Simply add another type of `m_cache_decrypt` with the corresponding signature: `typedef std::funtion<void(char* dst, const char* src, size_t size)> CacheDecryptChar;
2. Add yet another constructor of `ModelDeserializer` taking this second type of decryption functor
```cpp
ModelDeserializer(std::istream& model, ModelBuilder fn, CacheDecryptChar encrypt_fn);
```
3. The pointers may be stored in a union:
```cpp
union {
    CacheDecrypt m_cache_decrypt;
    CacheDecryptChar m_cache_char_decrypt;
}
bool decript_from_string = true; //default value otherwise use m_cache_char_decrypt;
```
This will provide explicit and understandable (as the result more reliable) solution.

Reply:
Done